drivers/can/can_stm32fd.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2020 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/can.h>
+#include <kernel.h>
+#include <soc.h>
+#include <stm32_ll_rcc.h>
+#include "can_stm32fd.h"
+#include <pinmux/stm32/pinmux_stm32.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
+
+#if CONFIG_CAN_STM32_CLOCK_DIVISOR != 1
+#if CONFIG_CAN_STM32_CLOCK_DIVISOR & 0x01
+#error CAN_STM32_CLOCK_DIVISOR invalid.\
+Allowed values are 1 or 2 * n, where n <= 15
+#endif
+#endif
+
+#define DT_DRV_COMPAT st_stm32_fdcan
+
+int can_stm32fd_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	*rate = LL_RCC_GetFDCANClockFreq(LL_RCC_FDCAN_CLKSOURCE) /
+		CONFIG_CAN_STM32_CLOCK_DIVISOR;
+	return 0;
+}
```

Comment:
```suggestion
int can_stm32fd_get_core_clock(const struct device *dev, uint32_t *rate)
{
	ARG_UNUSED(dev);
	*rate = LL_RCC_GetFDCANClockFreq(LL_RCC_FDCAN_CLKSOURCE);

	if (LL_RCC_PERIPH_FREQUENCY_NO == *rate) {
		LOG_ERR("Oscillator is not ready");
		return -EINVAL;
	}

	*rate /= CONFIG_CAN_STM32_CLOCK_DIVISOR;

	return 0;
}
```

The function `LL_RCC_GetFDCANClockFreq` has return values.